tensorboard/webapp/metrics/store/time_selection_utils.ts

```diff
@@ -0,0 +1,92 @@
+/* Copyright 2023 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+import {MinMaxStep, TimeSelection} from '../types';
+
+/**
+ * Ensures that value is within min max. If it is not, return the closest value that is.
+ * @param value
+ * @param min
+ * @param max
+ */
+export function clipStepWithinMinMax(value: number, min: number, max: number) {
+  if (value < min) {
+    return min;
+  }
+  if (value > max) {
+    return max;
+  }
+  return value;
+}
+
+/**
+ * Clips both start and end step of a time selection within min and max
+ * @param timeSelection
+ * @param param1
+ */
+export function maybeClipTimeSelection(
+  timeSelection: TimeSelection,
+  {minStep, maxStep}: MinMaxStep
+) {
+  const start = {
+    step: clipStepWithinMinMax(timeSelection.start.step, minStep, maxStep),
+  };
+  const end = timeSelection.end
+    ? {
+        step: clipStepWithinMinMax(timeSelection.end.step, minStep, maxStep),
+      }
+    : null;
+  return {
+    start,
+    end,
+  };
+}
+
+/**
+ * Removes the end step of a time selectio nif range selection is not enabled
+ * @param timeSelection
+ * @param rangeSelectionEnabled
+ */
+function maybeOmitTimeSelectionEnd(
```

Comment:
Could you describe what circumstance this might be used? Why end might not be null but decided as null afterward?

Reply:
I think my draft pr #6230 (it's note quite ready for review) does a decent job of explaining the problem I'm trying to solve here. You should be able to see all of this in action in the changes I made to metrics selectors there though.

In short the issue is that we don't want to actually remove the end time when range selection is marked as disabled. This ensures that re-enabling range selection results in the previous value.

Reply:
Thanks #6230 does help me understand what you're about to utilize formatTS function and  move all the logics from view. 

I agree we read from _redux to ui, not the other way around. But I don't know if this is about that. 
what I'm not sure is, I think we should let the view handles the view instead of moving those logic into selector, if I'm reading #6230 correctly. (at least formatTimeSelection() does look like updating the view)
Let's say user zoom in in linked-time, the time selection of this particular card isn't changed but the view is. but in 6230 it seems that part will be handled in selector.

What do you think? Do you feel strongly that we should handle this in selector? this might be a better of triggering Angular re-rendering with extra action and reducer handling?
If what I thought about what getCardTS selector is doing is wrong, it's actually not manipulating ui but state, I'm happy to go this way.

> we don't want to actually remove the end time when range selection is marked as disabled
Why is that? If range is disabled under linked time enabled, I think it's okay that end step is gone. Under step selector is even not an issue.

Reply:
So from your perspective the card should read the time selection from state, pipe it through `formatTimeSelection`, then pass it down to child components, right?

I'm inclined to move it to the selector layer to avoid needing to pass the value through the template.